pkg/agent/agent.go

```diff
@@ -29,21 +34,231 @@ import (
 	"github.com/nimbess/nimbess-agent/pkg/proto/cni"
 )
 
+// CNI Reply Values
+const (
+	CniOk                = 0
+	CniIncompatible      = 1
+	InvalidNetworkConfig = 2
+	ContainerNotExist    = 3
+	CniTryLater          = 11
+	DriverFailure        = 101
+)
+
+var ipAddr = net.IP{40, 0, 0, 0}
+
 // NimbessAgent represents the agent runtime server.
 // It contains a loadable runtime data plane driver to manage the data plane.
 // It includes a mutex used to handle locking between driver and agent events to force a
 // single-processed event pipeline.
 type NimbessAgent struct {
+	ID     uuid.UUID
 	Mu     *sync.Mutex
 	Config *NimbessConfig
 	Driver drivers.Driver
+	// Pipelines contains a map of port name to pipeline pointer
+	Pipelines map[string]*NimbessPipeline
+	// MetaPipelines contains a map of network/VRF name to abstract pipelines
+	MetaPipelines map[string]*NimbessPipeline
 }
 
 // Add implements CNI Add Handler.
 // It returns a CNI Reply to be sent to the Nimbess CNI client.
 func (s *NimbessAgent) Add(ctx context.Context, req *cni.CNIRequest) (*cni.CNIReply, error) {
-	//TODO implement
-	return &cni.CNIReply{}, nil
+	// Validate/parse CNI req
+	// TODO(TROZET)
+
+	// Hardcode to Kernel veth port for now
+	// Need to make this port name unique here, so use short NS name and port name
+	netNsSlice := strings.Split(req.NetworkNamespace, "/")
+	netNs := netNsSlice[len(netNsSlice)-1]
+	portName := fmt.Sprintf("port_%s_%s", netNs, req.InterfaceName)
+	log.Infof("Received port add request for: %s", portName)
+	port := &network.Port{
+		PortName:   portName,
+		Virtual:    true,
+		DPDK:       false,
+		IfaceName:  req.GetInterfaceName(),
+		NamesSpace: req.GetNetworkNamespace(),
+	}
+	// Check if this port already has a pipeline
+	if _, ok := s.Pipelines[portName]; ok {
+		log.Error("Pod already exists, invalid CNI ADD call")
+		return &cni.CNIReply{Result: CniIncompatible}, errors.New("pod already exists, invalid CNI ADD call")
+	}
+	// Call External IPAM
+	// TODO FIXME hack this for now for testing
+	ipAddr[3]++
+	port.IPAddr = fmt.Sprintf("%s/24", ipAddr.String())
+
+	// Initialize pipeline
+	// Check if meta pipeline exists for this network/VRF
```

Comment:
network/VRF implies network is VRF as opposed to network vrf - if I read this correctly. Is that the case? Or should we just say network?

Reply:
I should just say network. I don't think we need to support vrf yet.